Initial Python 3 support.

eisensheng opened the pull request:
Provide initial Python 3 support by using six where there are differences in the code between Python2 and Python3 behavior.

eisensheng committed:
Initial Python 3 support.

Tests are green.

mbertheau commented on requirements.in:
```diff
@@ -1,7 +1,9 @@
+
 # lowest supported django version
-Django<1.8
+Django>1.8
```
Can you keep it compatible with Django-1.8?

eisensheng replied:
Unfortunately not. Only the Django 1.9 line is compatible with Python 3.5 which is being used in the signup service. Anything below 1.9 throws up during the test run.

```
  File ".../.tox/py35/lib/python3.5/site-packages/django/utils/html_parser.py", line 12, in <module>
    HTMLParseError = _html_parser.HTMLParseError
AttributeError: module 'html.parser' has no attribute 'HTMLParseError'
```

I changed the version requirements though to be a little bit more strict.

mbertheau replied:
That seriously decreases the value of the work for us. Django 1.9 is supported until 8 months from now, whereas Django 1.8 is supported until 20 months from now.

Additionally the django documentation states here https://docs.djangoproject.com/en/1.8/releases/1.8/ that Python 3.5 is supported. In fact a problem with HTMLParseError is mentioned here https://code.djangoproject.com/ticket/23763 and has been fixed here https://github.com/django/django/commit/b07aa52e8a8e4c7fdc7265f75ce2e7992e657ae9 and the fix should be included in the latest 1.8 release. Please check that again.

Is there anything else preventing us from using Django 1.8?

eisensheng replied:
Interesting. I was pretty sure that I tested it with the 1.8 line of Django and it didn't work either. I've explicitly pinned the 1.8.14 version and upwards together with 1.9.8 and upwards until 1.10 line for now and it works magically. :sparkles: 

The test suite will be run by tox for the Django 1.8 line and 1.9 line under Python 2.7 and 3.5 now. It was necessary to split the requirements to achieve this though. Pip doesn't like it that much when a dependency is specified twice it seems.

mbertheau replied:
We can drop Django 1.9 support.

eisensheng committed:
Support testing with Django1.8,1.9 with Py27,Py35

Split requirements in testing and runtime requirements to support
running the test suite with latest Django 1.8 and 1.9 line and Python in
the version 2.7 and 3.5

This is necessary because pip doesn't like it when Django is specified as
a dependency twice in a requirements file and through the tox deps
configuration parameter.

eisensheng committed:
Fix compatibility issue regarding slice builtin

eisensheng committed:
Support wildcard in whitelisting

This is useful to add any address under the example.test domain for example.

eisensheng committed:
Add missing slash for API endpoint

Azd325 commented:
Has already Python 3 support.
